### ISSUE TITLE
multiple links of the same relation type

### ISSUE THREAD

**Issue by dret** [smizell/restfuljson#8]
just looking ad the example of `author_url` it seems like the implicit suggestion is to encode the link relation type as a prefix of the `_url` naming scheme. if that's the case, then the obvious question raised by the example would be how to apply the pattern for more than one author? making the member value an array would be against the pattern, but addressing this and giving some guidance might be helpful.

**Comment by fosrias** [comment]
@dret Interesting point, so you are implying the need for singular and plural suffixes, e.g. `_urls` if we wanted to include a list of urls. Albeit, this can also be accomplished (although not necessarily optimally) in current with:

```javascript
"authors": [
  { "url": "..." },
  { "url": "..." }
```

**Comment by smizell** [comment]
@dret I have some plans around a project called "RESTful Patterns" to describe some best practices there (along with other concepts). Ideally, any implementation is domain specific, so there are actually a few ways that people can provide a collection of links and it's up to them. As @fosrias showed above, an API can do a normal `authors` property with an array of author objects. However, I would foresee that people need more control than this, so they'll want to follow a collection pattern. This would let people either link to the authors:

```
{
    "authors_url": "..."
}
```

Or embed that collection in another representation.

```
{
    "authors": {
        "items": [
            {"url": "..."},
            {"url": "..."},
            {"url": "..."}
        ]
    }
}
```

Long story short, I do have some plans to provide patterns around this, but it's in a separate project.

**Comment by smizell** [comment]
Just to add, if you do this, clients can be instructed through the documentation to look for either an `authors_url` for a link or `authors` for the embedded collection.

**Comment by dret** [comment]
On Sep 7, 2017, at 09:51, Mark W. Foster <notifications@github.com> wrote:
> @dret Interesting point, so you are implying the need for singular and plural suffixes, e.g. _urls if we wanted to include a list of urls. 
> 

i did not really want to imply how to do it. i just wanted to point out that having more than one link of a given type is a common pattern in many APIs, and that it might be helpful to provide guidance for this pattern.

**Comment by smizell** [comment]
Yes, I do think that's a good point. Thanks for providing this. We may find an FAQ or additional guidelines may help within the context of this doc.

@dret any other guidelines you think may help would be great! :)

**Comment by dret** [comment]
On 2017-09-08 21:54, Stephen Mizell wrote:
> @dret <https://github.com/dret> any other guidelines you think may help 
> would be great! :)

as mentioned on twitter to me this looks too minimal. for linking to be 
meaningful you need affordances, just having URIs usually doesn't cut 
it. https://github.com/dret/hyperpedia/blob/master/concepts.md is an 
abstract list i came up with a while ago. the current work on 
https://tools.ietf.org/html/draft-wilde-linkset-link-rel is a more 
concrete list and representation of how to do (some of) this. i think 
there are different ways in which you can approach this (and different 
people have different preferences), but calling it "RESTful" and having 
no model for affordances seems a bit odd to me.

**Comment by smizell** [comment]
I've moved this conversation to #9 to keep it separate. If you have time @dret I'd love to continue it.

**Comment by smizell** [comment]
After thinking about it and tinkering with it, I do think some guidance for this would be helpful. Thanks for you idea and patience @dret :) I'd like to work through the options I see here.

My personal favorite is adding `_urls` which is an array of URIs. We actually had this in the beginning and then moved away from it. The downside is that I was unable to find this pattern used in any of the APIs we have listed, so it's a slight deviation of expressing an already-existing pattern for hypermedia.

Next on the list for me would be to instruct producers to simply to just give an array of URIs. This is what Django REST Framework does with their [hyperlinked field](http://www.django-rest-framework.org/api-guide/relations/#hyperlinkedrelatedfield). This is more of a guideline than anything. It's limit is that a consumer cannot find the links programmatically.

Lastly, we could guide people to use partial objects. It would be an array of objects with a `url` field and that's it. Consumers would decide if they want to request the link to get the full representation. The biggest issue here is that there is no indication it is partial other than the object has just one field `url`. This isn't terrible, but it moves a little beyond the simplicity that's captured.

@fosrias @dret thoughts? Any option feel better than the others?
